File: src/Presentation/Hubs/GameCreatorHub.cs
using Application.Common.Models;
using Application.MatchListItems.Commands;
using Application.MatchListItems.Queries;
using MediatR;
using Microsoft.AspNetCore.SignalR;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Presentation.Hubs
{
    public class GameCreatorHub : Hub
    {
        readonly IMediator _mediator;

        public GameCreatorHub(IMediator mediator)
        {
            _mediator = mediator;
        }

        public async Task GetMatchListItems()
        {
            await _mediator.Send(new RemoveEmptyMatchesCommand());

            List<MatchListItem> matchListItems = await _mediator.Send(new GetAllMatchListItemsQuery());

            await Clients.Client(Context.ConnectionId).SendAsync("RecieveAndRenderListOfMatches", matchListItems);
        }

        public async Task TryJoinMatch(string id, string password)
        {
            MatchListItem match = await _mediator.Send(new GetMatchByIdQuery { Id = id });

            bool isPasswordCorrect = match.Password == password;
            if (isPasswordCorrect)
            {
                await Clients.Caller.SendAsync("RecieveMatchUrl", match.Url);
                return;
            }

            await Clients.Caller.SendAsync("SendNotificationAboutIncorrectPassword");
        }

        public async Task CreateMatch(MatchListItem match)
        {
            List<MatchListItem> matchListItems = await _mediator.Send(new GetAllMatchListItemsQuery());
            bool matchDoesNotExist = matchListItems.FirstOrDefault(x => x.Id == match.Id) == null;

            if (matchDoesNotExist)
            {
                _mediator.Send(new AddMatchListItemCommand { MatchListItem = match });
            }
        }
    }
}

File: src/Presentation/scripts/gameCreatorHub.ts
﻿declare var signalR: any;
import * as indexView from './views/homeView.js';
import { MatchListItem } from './models/MatchListItem.js';
import { navigateToMatchUrl } from './home.js';
import * as notificationSender from './views/notificationSender.js';


var hub = new signalR.HubConnectionBuilder()
    .withUrl('/gameCreatorHub')
    .build();


hub.on('SendNotificationAboutIncorrectPassword', (): void => {
    notificationSender.sendNotificationAboutIncorrectPassword();
});

hub.on('RecieveMatchUrl', (url: string): void => {
    navigateToMatchUrl(url);
});

hub.on('RecieveAndRenderListOfMatches', (matches: Array<MatchListItem>) => {
    indexView.renderMatchesInMatchList(matches);
});

hub.start().then(function () {
}).catch(function (err) {
    return console.error(err.toString());
});


export const createMatch = async (match: MatchListItem): Promise<void> => {
    await hub.invoke('CreateMatch', match);
};

export const refreshListOfGames = async (): Promise<void> => {
    await hub.invoke('GetMatchListItems');
};

export const tryJoinMatch = async (id: string, password: string): Promise<void> => {
    await hub.invoke('TryJoinMatch', id, password)
};
